Source/Core/VideoBackends/Metal/MTLMain.mm

```diff
@@ -0,0 +1,144 @@
+    // so we have to check it ourselves.
+    size_t selected_adapter_index = static_cast<size_t>(g_Config.iAdapter);
+    if (selected_adapter_index >= devs.size())
+    {
+      WARN_LOG_FMT(VIDEO, "Metal adapter index out of range, selecting default adapter.");
+      selected_adapter_index = 0;
+    }
+    MRCOwned<id<MTLDevice>> adapter = std::move(devs[selected_adapter_index]);
+    Util::PopulateBackendInfoFeatures(&g_Config, adapter);
+
+    // With the backend information populated, we can now initialize videocommon.
+    InitializeShared();
+
+    MRCOwned<CAMetalLayer*> layer = MRCRetain(static_cast<CAMetalLayer*>(wsi.render_surface));
+    [layer setDevice:adapter];
+    if (Util::ToAbstract([layer pixelFormat]) == AbstractTextureFormat::Undefined)
+      [layer setPixelFormat:MTLPixelFormatBGRA8Unorm];
+    CGSize size = [layer bounds].size;
+    float scale = [layer contentsScale];
+
+    ObjectCache::Initialize(std::move(adapter));
+    g_state_tracker = std::make_unique<StateTracker>();
+    g_renderer = std::make_unique<Renderer>(std::move(layer), size.width * scale,
+                                            size.height * scale, scale);
+    g_vertex_manager = std::make_unique<VertexManager>();
+    g_perf_query = std::make_unique<PerfQuery>();
+    g_framebuffer_manager = std::make_unique<FramebufferManager>();
+    g_texture_cache = std::make_unique<TextureCacheBase>();
+    g_shader_cache = std::make_unique<VideoCommon::ShaderCache>();
+
+    if (!g_vertex_manager->Initialize() || !g_shader_cache->Initialize() ||
+        !g_renderer->Initialize() || !g_framebuffer_manager->Initialize() ||
+        !g_texture_cache->Initialize())
+    {
+      PanicAlertFmt("Failed to initialize renderer classes");
+      Shutdown();
+      return false;
+    }
+
+    g_shader_cache->InitializeShaderCache();
+
+    return true;
+  }
+}
+
+void Metal::VideoBackend::Shutdown()
+{
+  g_shader_cache->Shutdown();
+  g_renderer->Shutdown();
+
+  g_shader_cache.reset();
+  g_texture_cache.reset();
+  g_framebuffer_manager.reset();
+  g_perf_query.reset();
+  g_vertex_manager.reset();
+  g_renderer.reset();
+  g_state_tracker.reset();
+  ObjectCache::Shutdown();
+  ShutdownShared();
+}
+
+void Metal::VideoBackend::InitBackendInfo()
+{
+  @autoreleasepool
+  {
+    Util::PopulateBackendInfo(&g_Config);
+    auto adapters = Util::GetAdapterList();
+    Util::PopulateBackendInfoAdapters(&g_Config, adapters);
+    if (!adapters.empty())
+    {
+      // Use the selected adapter, or the first to fill features.
+      size_t index = static_cast<size_t>(g_Config.iAdapter);
+      if (index >= adapters.size())
+        index = 0;
+      Util::PopulateBackendInfoFeatures(&g_Config, adapters[index]);
+    }
+  }
+}
+
+void Metal::VideoBackend::PrepareWindow(WindowSystemInfo& wsi)
+{
+  if (wsi.type != WindowSystemType::MacOS)
+    return;
+  NSView* view = static_cast<NSView*>(wsi.render_surface);
```

Comment:
In my iOS fork, I pass the `CAMetalLayer*` in `render_surface`.

Reply:
In that case, we could leave the `wsi.type != WindowSystemType::MacOS` as is, and on iOS it would just leave the CAMetalLayer there.

Reply:
Sounds good to me.